// @smartspectra/node-sdk/js/constants.js
// constants.js
// Copyright (C) 2026 Presage Technologies, Inc.
//
// SPDX-License-Identifier: LicenseRef-Proprietary
//
// Pure-value exports — public enums and metric bundles. No Node-only
// dependencies so this module can be bundled into a browser renderer
// without dragging in koffi, fs, or path.
//
// Integer values must match the SmartSpectra C++ headers and
// metric_types.proto wire values; stable across SDK versions.

'use strict';

const ProcessingStatus = Object.freeze({
    kUninitialized: 0,
    kIdle:          1,
    kStarting:      2,
    kRunning:       3,
    kStopping:      4,
    kError:         5,
});

const PixelFormat = Object.freeze({
    kRGB:  0,
    kBGR:  1,
    kRGBA: 2,
    kBGRA: 3,
    kNV12: 4,
    kNV21: 5,
    kYUYV: 6,
});

const ValidationCode = Object.freeze({
    kOk:                 0,
    kNoFaceFound:        1,
    kMultipleFacesFound: 2,
    kFaceNotCentered:    3,
    kFaceSizeOutOfRange: 4,
    kTooDark:            5,
    kTooBright:          6,
    kChestNotVisible:    7,
    kCameraTuning:       10,
    kFrameRateTooLow:    11,
});

const SmartSpectraErrorCode = Object.freeze({
    kOk:                    0,
    kInvalidState:          1,
    kAuthenticationFailed:  2,
    kConfigurationFailed:   3,
    kCreditExhausted:       4,
    kNetworkError:          5,
    kServerError:           6,
    kInputUnavailable:      7,
    kProcessingFailed:      8,
    kFrameConversionFailed: 9,
    kNonMonotonicTimestamp: 10,
    kTimestampGap:          11,
});

// Spatial transform applied to each pushed frame before it enters the
// graph. Pass to `useCustomInput(frameTransform)`.
const FrameTransform = Object.freeze({
    kNone:             0,
    kRotate90CW:       1,
    kRotate90CCW:      2,
    kRotate180:        3,
    kMirrorHorizontal: 4,
    kMirrorVertical:   5,
});

// MetricType integer codes grouped by measurement type. Spread into
// `requestedMetrics`; omitting `requestedMetrics` is equivalent to passing
// `breathingMetrics`.
const breathingMetrics = Object.freeze([
    0,  // CHEST_BREATHING
    1,  // ABDOMEN_BREATHING
    2,  // BREATHING_RATE
    3,  // BREATHING_AMPLITUDE
    4,  // APNEA
    5,  // RESPIRATORY_LINE_LENGTH
    6,  // BASELINE
    7,  // INHALE_EXHALE_RATIO
]);

const cardioMetrics = Object.freeze([
    15, // PULSE_RATE
    16, // ARTERIAL_PRESSURE_TRACE
    17, // HRV
]);

const faceMetrics = Object.freeze([
    11, // FACE_LANDMARKS
    12, // BLINKING
    13, // TALKING
    14, // EXPRESSIONS
]);

const micromotionMetrics = Object.freeze([
    8,  // GLUTES_MICROMOTION
    9,  // KNEES_MICROMOTION
]);

const edaMetrics = Object.freeze([
    10, // EDA_TRACE
]);

// The default metric bundle used when `requestedMetrics` is omitted; currently
// equals the breathing bundle.
const defaultSupportedMetrics = breathingMetrics;

module.exports = {
    ProcessingStatus,
    PixelFormat,
    ValidationCode,
    SmartSpectraErrorCode,
    FrameTransform,
    breathingMetrics,
    cardioMetrics,
    faceMetrics,
    micromotionMetrics,
    edaMetrics,
    defaultSupportedMetrics,
};
