// @smartspectra/node-sdk/js/resolve-native.js
// resolve-native.js
// Copyright (C) 2026 Presage Technologies, Inc.
//
// SPDX-License-Identifier: LicenseRef-Proprietary
//
// Picks the right libsmartspectra_capi.{dll,dylib,so} for the current
// platform+arch and returns a path that koffi.load() can consume.
//
// The native runtime closure (shim + libsmartspectra + OpenCV + graph data,
// plus a Vulkan loader on Linux/Windows; on macOS MoltenVK is statically
// linked into libsmartspectra, so no separate Vulkan dylib ships) lives in
// per-platform packages — `@smartspectra/node-sdk-<plat>-<arch>`
// — declared as regular (HARD) dependencies of the main package. There is no
// postinstall download and no install script in the published tarball.
//
// Why hard deps and not optionalDependencies gated by os/cpu: GitLab's npm
// registry strips optionalDependencies/os/cpu from the packument, so optional
// gating never resolves. As hard deps every platform installs on every host,
// and this resolver picks the one matching process.platform/process.arch. See
// .session-logs/nodejs-bundle-all-platforms.md and package.json#_nativeRuntime_doc.
//
// Resolution order:
//   1. SMARTSPECTRA_CAPI_PATH env var — full path to the shared library.
//      The sole dev/override escape hatch (e.g. running against a local
//      source build, or a custom-bundled closure). In a source checkout the
//      platform packages are NOT installed, so dev uses this override.
//   2. The host platform package's root, located via require.resolve. The shim
//      sits at the package root; the rest of the closure sits beside it with
//      install_names / RPATHs pre-rewritten so it loads with zero env vars.
//
// If neither resolves, we throw a single clear error naming the missing
// platform package and the SMARTSPECTRA_CAPI_PATH override.

'use strict';

const fs   = require('fs');
const path = require('path');

const LIB_BASENAME = {
    darwin: 'libsmartspectra_capi.dylib',
    linux:  'libsmartspectra_capi.so',
    win32:  'smartspectra_capi.dll',
};

// `<platform>-<arch>` — matches process.platform / process.arch and the
// per-platform package suffix.
function platformArch() {
    return `${process.platform}-${process.arch}`;
}

function platformPackageName() {
    return `@smartspectra/node-sdk-${platformArch()}`;
}

function resolveNativeLibrary() {
    const libName = LIB_BASENAME[process.platform];
    if (!libName) {
        throw new Error(
            `@smartspectra/node-sdk: unsupported platform ${process.platform}. ` +
            `Supported: darwin (macOS), linux, win32 (Windows).`);
    }

    const override = process.env.SMARTSPECTRA_CAPI_PATH;
    if (override) {
        // nosemgrep: detect-non-literal-fs-filename - safe: path is the consumer-set SMARTSPECTRA_CAPI_PATH override, same trust boundary as the process; no attacker input.
        if (!fs.existsSync(override)) {
            throw new Error(
                `@smartspectra/node-sdk: SMARTSPECTRA_CAPI_PATH=${override} ` +
                `points at a file that does not exist.`);
        }
        return override;
    }

    const platformPkg = platformPackageName();
    let pkgRoot;
    try {
        // Resolve the package's own package.json, then take its directory —
        // require.resolve(pkgName) would need a "main"/"exports" entry the
        // binary-only package intentionally lacks.
        pkgRoot = path.dirname(require.resolve(`${platformPkg}/package.json`));
    } catch (e) {
        // A corrupt install (broken symlink, permissions) surfaces as something
        // other than a missing module — don't bury its real cause under the
        // "not installed" guidance, which would send the user down the wrong path.
        if (e && e.code !== 'MODULE_NOT_FOUND') {
            throw new Error(
                `@smartspectra/node-sdk: failed to resolve the native runtime package ` +
                `"${platformPkg}" (${e.code || 'unknown error'}): ${e.message}`);
        }
        throw new Error(
            `@smartspectra/node-sdk: the native runtime package "${platformPkg}" is not installed.\n` +
            `It is a regular dependency of @smartspectra/node-sdk, so this usually means your ` +
            `platform (${platformArch()}) is unsupported, or the install was incomplete ` +
            `(e.g. \`npm install --omit=optional\` does NOT skip it, but a partial/offline ` +
            `install might).\n` +
            `Supported platforms: darwin-arm64, linux-x64, linux-arm64, win32-x64 ` +
            `(glibc only — musl/Alpine Linux is not supported).\n` +
            `To point at a custom build, set SMARTSPECTRA_CAPI_PATH to the shared library path.`);
    }

    const libPath = path.join(pkgRoot, libName);
    // nosemgrep: detect-non-literal-fs-filename - safe: path is derived from require.resolve of the installed platform package plus a fixed library basename; same trust boundary as the process, no attacker input.
    if (!fs.existsSync(libPath)) {
        throw new Error(
            `@smartspectra/node-sdk: "${platformPkg}" is installed but ${libName} is missing from it ` +
            `(${libPath}). The package may be corrupt — try reinstalling.`);
    }
    return libPath;
}

module.exports = {
    resolveNativeLibrary,
    // Exposed for diagnostics and tests.
    LIB_BASENAME,
    platformArch,
    platformPackageName,
};
