// @smartspectra/node-sdk/js/smartspectra.js
// smartspectra.js
// Copyright (C) 2026 Presage Technologies, Inc.
//
// SPDX-License-Identifier: LicenseRef-Proprietary

'use strict';

const { Session } = require('./ffi');
const {
    ProcessingStatus,
    PixelFormat,
    ValidationCode,
    SmartSpectraErrorCode,
    FrameTransform,
    breathingMetrics,
    cardioMetrics,
    faceMetrics,
    micromotionMetrics,
    edaMetrics,
    defaultSupportedMetrics,
} = require('./constants');
const { version: PACKAGE_VERSION } = require('../package.json');

const SUPPORTED_EVENTS = new Set([
    'processingStatus',
    'validationStatus',
    'metrics',
    'accumulatedMetrics',
    'insight',
    'error',
    'frameSentThrough',
    'videoOutput',
]);

// FrameTransform.kNone — kept here so the constructor doesn't depend on the
// FrameTransform object declared later.
const DEFAULT_FRAME_TRANSFORM = 0;

class SmartSpectraSDK {
    /** SDK package version (matches `package.json#version`). */
    static get version() { return PACKAGE_VERSION; }

    /**
     * @param {Object} [options]
     * @param {string} [options.apiKey]
     * @param {number[]} [options.requestedMetrics]
     *   MetricType integer codes. Defaults to `breathingMetrics`.
     * @param {boolean} [options.enableAccumulatedOutput=false]
     */
    constructor(options = {}) {
        this._options = {
            apiKey:                  options.apiKey ?? '',
            requestedMetrics:        options.requestedMetrics,
            enableAccumulatedOutput: !!options.enableAccumulatedOutput,
        };
        this._listeners = new Map();
        // Pending input source set by useCustomInput()/useCamera()/useFile() and
        // consumed by start(). Shape: { kind: 'custom'|'camera'|'file', ... }.
        // The last use* call wins.
        this._source = null;
        this._session = null;  // lazy: created on start()
    }

    // ---------- public API --------------------------------------------------

    /**
     * Register a callback for a named event. Returns `this` for chaining.
     * One listener per event; re-registration replaces.
     */
    on(event, callback) {
        if (typeof event !== 'string' || typeof callback !== 'function') {
            throw new TypeError('on(event: string, callback: Function)');
        }
        if (!SUPPORTED_EVENTS.has(event)) {
            throw new RangeError(`on(): unsupported event ${event}`);
        }
        this._listeners.set(event, callback);
        return this;
    }

    /**
     * Select the custom frame-push input source (you push frames via
     * sendFrame() after start()). Returns `this` for chaining.
     *
     * @param {number} [frameTransform=FrameTransform.kNone]
     *   Spatial transform applied to every pushed frame.
     */
    useCustomInput(frameTransform = DEFAULT_FRAME_TRANSFORM) {
        this._source = { kind: 'custom', frameTransform: frameTransform | 0 };
        return this;
    }

    /**
     * Select a live camera as the input source. The SDK opens the camera and
     * pumps frames internally on start() — no sendFrame() needed. Returns
     * `this` for chaining.
     *
     * @param {Object} [options] { deviceIndex?, width?, height?, fps?, frameTransform? }
     *
     * Captures from a camera in THIS process (headless / Node use). In an
     * Electron app the camera is normally owned by the renderer via
     * getUserMedia — use the renderer SDK's useMediaStream() there instead.
     */
    useCamera(options = null) {
        this._source = { kind: 'camera', options };
        return this;
    }

    /**
     * Select a pre-recorded video file as the input source. start() begins
     * playback (non-blocking — it runs on SDK worker threads); use
     * waitUntilComplete() or watch the `'processingStatus'` event for the idle
     * transition to detect end-of-file. Returns `this` for chaining.
     *
     * @param {string} videoPath
     * @param {Object} [options]
     *   { timestampsPath?, interframeDelayMs?, startOffsetMs?, maxDurationMs?, frameTransform? }
     */
    useFile(videoPath, options = null) {
        if (typeof videoPath !== 'string' || !videoPath) {
            throw new TypeError('useFile: videoPath must be a non-empty string');
        }
        this._source = { kind: 'file', videoPath, options };
        return this;
    }

    /**
     * Initialize the session and start the configured input source. Returns
     * once running — for custom input the graph is armed for sendFrame(); for
     * camera/file the SDK pumps frames itself. Configure a source with
     * useCustomInput() / useCamera() / useFile() first.
     */
    start() {
        const source = this._source;
        if (!source) {
            throw new Error(
                'start(): no input source configured. Call useCustomInput(), ' +
                'useCamera(), or useFile() first.');
        }
        this._ensureSession();
        switch (source.kind) {
            case 'custom':
                this._session.startCustom(source.frameTransform);
                break;
            case 'camera':
                this._session.startCamera(source.options);
                break;
            case 'file':
                this._session.startFile(
                    source.videoPath,
                    (source.options && source.options.timestampsPath) || null,
                    source.options);
                break;
        }
    }

    stop() {
        if (!this._session) return;
        this._session.stop();
    }

    /**
     * Async variant of `stop()`. The underlying C++ stop blocks until the
     * processing pipeline drains — can take multiple seconds. Prefer this in
     * event-loop-sensitive contexts (e.g. the Electron main process serving
     * IPC) so the runtime stays responsive.
     */
    stopAsync() {
        if (!this._session) return Promise.resolve();
        return this._session.stopAsync();
    }

    /**
     * Recover from a kError state. Rebuilds the internal graph. The input
     * source is discarded — call useCustomInput() / useCamera() / useFile()
     * AND start() again to drive new input.
     */
    reset() {
        if (!this._session) return;
        this._session.reset();
        this._source = null;
    }

    /**
     * Block until the session reaches a terminal state (idle or error), or the
     * timeout elapses. Returns true if the session settled, false on timeout.
     * `timeoutMs <= 0` (the default) waits indefinitely. With no active session
     * there is nothing to wait for, so it returns true immediately.
     *
     * Useful after useFile() + start() to block until end-of-file.
     */
    waitUntilComplete(timeoutMs = 0) {
        if (!this._session) return true;
        return this._session.wait(timeoutMs);
    }

    /**
     * Dispatch an on-demand insight prompt against the running session.
     * Returns the request id; the matching Insight response is delivered
     * asynchronously through the `'insight'` event with the same id.
     */
    requestInsight(text) {
        if (typeof text !== 'string') {
            throw new TypeError('requestInsight: text must be a string');
        }
        this._ensureSession();
        return this._session.requestInsight(text);
    }

    /** Current ProcessingStatus integer value (kUninitialized=0..kError=5). */
    get processingStatus() {
        return this._session ? this._session.getStatus() : 0;
    }

    /**
     * Push a raw video frame. Requires useCustomInput() + start() first.
     * Returns true on success; throws on input-validation or SDK-level
     * failure.
     */
    sendFrame(buffer, width, height, strideBytes, pixelFormat, timestampUs) {
        if (!this._session || !this._source || this._source.kind !== 'custom') {
            throw new Error('sendFrame(): call useCustomInput() + start() first');
        }
        return this._session.sendFrame(
            buffer, width, height, strideBytes, pixelFormat, timestampUs);
    }

    /**
     * Tear down the session. Idempotent. Always call on shutdown so the SDK
     * worker threads join cleanly before the koffi trampolines are released.
     * Returns a promise that resolves once native teardown completes (a
     * stopAsync() in flight defers it); await it before standing up a
     * replacement session, since SDK state is process-global.
     */
    destroy() {
        if (!this._session) return Promise.resolve();
        const teardown = this._session.destroy();
        this._session = null;
        this._source = null;
        return teardown;
    }

    // ---------- internal ----------------------------------------------------

    _ensureSession() {
        if (this._session) return;
        // Listeners may be added/replaced after start(), so the lookup
        // happens at emit time rather than capture time.
        this._session = new Session(this._options, {
            onStatus: (status) => this._emit('processingStatus', status),
            onValidation: (code, hint, ts) =>
                this._emit('validationStatus', code, Number(ts), hint),
            onMetrics: (buf, ts) => this._emit('metrics', buf, Number(ts)),
            onError: (code, message, retryable) =>
                this._emit('error', code, message, retryable),
            onFrame: (sent, ts) =>
                this._emit('frameSentThrough', sent, Number(ts)),
            onAccumulatedMetrics: (buf, ts) =>
                this._emit('accumulatedMetrics', buf, Number(ts)),
            onInsight: (buf, requestId) =>
                this._emit('insight', buf, requestId),
            onVideoOutput: (buf, width, height, stride, pixelFormat, ts) =>
                this._emit('videoOutput', buf, width, height, stride, pixelFormat, Number(ts)),
        });
    }

    _emit(event, ...args) {
        const cb = this._listeners.get(event);
        if (!cb) return;
        try {
            cb(...args);
        } catch (e) {
            // Isolate listener exceptions so a user-code throw on a
            // worker-thread callback doesn't tear down the SDK.
            console.error(`SmartSpectraSDK: listener for '${event}' threw:`, e);
        }
    }
}

module.exports = {
    SmartSpectraSDK,
    ProcessingStatus,
    PixelFormat,
    ValidationCode,
    SmartSpectraErrorCode,
    FrameTransform,
    breathingMetrics,
    cardioMetrics,
    faceMetrics,
    micromotionMetrics,
    edaMetrics,
    defaultSupportedMetrics,
};
